hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPServiceActor.java

```diff
@@ -791,6 +792,9 @@ private void offerService() throws Exception {
           shouldServiceRun = false;
           return;
         }
+        if (InvalidBlockReportLeaseException.class.getName().equals(reClass)) {
+          fullBlockReportLeaseId = 0;
```

Comment:
@sodonnel Thanks for your works here. Do we also need set `forceFullBlockReport` to true here? Otherwise, datanode will send block report at next 6 hour by default, right?

Reply:
At line 717, we can see where it attempts to get a lease from the heartbeat if the lease in the DN == 0:

```
   boolean requestBlockReportLease = (fullBlockReportLeaseId == 0) &&
                  scheduler.isBlockReportDue(startTime);
```

So its the isBlockReportDue that controls this. Then later, if we have a non zero lease, it will try to create the block report:

```
        boolean forceFullBr =
            scheduler.forceFullBlockReport.getAndSet(false);
        if (forceFullBr) {
          LOG.info("Forcing a full block report to " + nnAddr);
        }
        if ((fullBlockReportLeaseId != 0) || forceFullBr) {
          cmds = blockReport(fullBlockReportLeaseId);
          fullBlockReportLeaseId = 0;
        }
```
Its really the `isBlockReportDue()` method that controls whether a new one should be sent of not, and that is based on time since the last one. The the `blockReport()`, it updates the time after a successful block report, but if it gets an exception, like this change causes, it will not update the time and so it will try again on the next heartbeat if it gets a new lease.

I think `forceFullBlockReport` is only for tests, or the command to force a DN block from the CLI.

Reply:
Also, do you have any idea about fixing the checkstyle issue? As I mentioned above, trying to add a package-info.java file broke my compile locally.

Reply:
As we don't expect to reach here frequently (hopefully datanode is able to acquire lease successfully most of the times), perhaps we can add one liner log to indicate that the particular block id went through this trouble?

Reply:
Or maybe `reportId` and `leaseId` can be added as constructor argument to `InvalidBlockReportLeaseException`. This way `RemoteException in offerService` log will likely print them anyways?

Reply:
> Its really the isBlockReportDue() method that controls whether a new one should be sent of not, and that is based on time since the last one. The the blockReport(), it updates the time after a successful block report, but if it gets an exception, like this change causes, it will not update the time and so it will try again on the next heartbeat if it gets a new lease.

Thanks for the detailed explain. Make sense to me.

> perhaps we can add one liner log to indicate that the particular FBR went through this trouble (i.e. log report id and lease id)

+1 from my side.

Reply:
I added the report and lease ID into the exception message, so it will get logged as part of the exception.

Still stuck on that checkstyle violation :-(